graphics/blender/Portfile

```diff
@@ -114,6 +116,46 @@ post-patch {
             ${platform_apple_xcode.cmake}
         reinplace -E {s/(OSX_MIN_DEPLOYMENT_TARGET) 10.13/\1 10.11/} \
             ${platform_apple_xcode.cmake}
+
+        # Starting in Blender 2.90, the devs moved to using C++17 to
+        # compile Cycles. However, on older versions of macOS, this
+        # causes the build to fail with the error 'Undefined
+        # symbols for architecture x86_64: "operator delete(void*,
+        # std::align_val_t)", referenced from:
+        # ccl::OSLRenderServices::~OSLRenderServices() in
+        # libcycles_kernel_osl.a(osl_services.cpp.o)'. The first few
+        # references below seemed to point towards something related to
+        # needing to disable -falign-new by using
+        # -DCOMPILER_HAS_F_ALIGNED_NEW=OFF, but it turns out that the
+        # error is actually being caused by the fact that the devs are
+        # adding -faligned-allocation in the file
+        # 'intern/cycles/kernel/osl/CMakeLists.txt'. It seems that the
+        # real problem is that Clang on older macOSes do not have
+        # aligned allocation when compiling using C++17. Thus, we need
+        # use -fno-aligned-allocation. However, we also cannot simply
+        # remove the section of code from 'CMakeLists.txt' that adds
+        # -faligned-allocation, because if we do, the build will fail
+        # with the warning/error: "aligned deallocation function of type
+        # 'void (void *, std::align_val_t) noexcept' is only available
+        # on macOS 10.14 or newer"; "note: if you supply your own
+        # aligned allocation functions, use -faligned-allocation to
+        # silence this diagnostic". Thus, in order to build
+        # successfully, we need to explicitly use
+        # -fno-aligned-allocation on older macOSes, and explicitly use
+        # -faligned-allocation on newer macOSes.
+        #
+        # References:
+        # * https://github.com/facebook/folly/issues/864
+        # * https://www.veripool.org/issues/1231
+        # * https://trac.macports.org/ticket/47191
+        # * https://bugs.llvm.org/show_bug.cgi?id=22951
+        # * https://stackoverflow.com/questions/31686508/why-is-delete-operator-required-for-virtual-destructors
+        # * https://bugs.freedesktop.org/show_bug.cgi?id=54948
+        # * https://github.com/pybind/python_example/issues/48
+        # * https://github.com/llvm-mirror/clang/blob/master/test/SemaCXX/unavailable_aligned_allocation.cpp
+        # * See also the Portfile for port:folly.
+        reinplace -E {s/(-f)(aligned-allocation)/\1no-\2/} \
+            ${worksrcpath}/intern/cycles/kernel/osl/CMakeLists.txt
```

Comment:
If the build requires c++17 you should change the requested standard to

compiler.cxx_standard 2017

To indicate this.

it would also be useful to know if you in addition  blacklist older macports clang versions, ie macports-clang-[5-9].0 which then will force the use of the most recent clang 11 version, if this then means the hadk here is no longer required ?

Reply:
B.t.w. You should also update the regex

regexp {clang-mp-([0-9]+(?:\.[0-9]+)*)} \
    ${configure.cc} result llvm_version

to work properly with macports clang 10 and newer, i.e. where the major version has double digits and there is no minor version, e.g. macports-clang-11.

Reply:
The entire build doesn't rely on C++17... only one specific piece, the Cycles renderer, requires C++17. In fact, there are portions which rely on C++11, and others which need C++14. With so many (nearly independent) moving parts, each with its own needs for C++ standard, it's kind of difficult to use `compiler.cxx_standard` to force a single C++ standard onto the entire source code. Rather, I've been using it as a sort of minimum... Blender uses newer C++ standards in the individual components' CMake files wherever it needs to.

> it would also be useful to know if you in addition blacklist older macports clang versions, ie macports-clang-[5-9].0 which then will force the use of the most recent clang 11 version, if this then means the hack here is no longer required ?

As far as I'm aware, I'm not blacklisting older macports clangs; I'm only blacklisting Xcode Clang. Could you point out which part of my portfile you are referring to?

> B.t.w. You should also update the regex
> ```
> regexp {clang-mp-([0-9]+(?:\.[0-9]+)*)} \
>     ${configure.cc} result llvm_version
> ```
> to work properly with macports clang 10 and newer, i.e. where the major version has double digits and there is no minor version, e.g. macports-clang-11.

I believe that regex should work with clang 10 and newer. The `(?:)*` makes the `.y` portion of version `x.y` optional.

Reply:
> The entire build doesn't rely on C++17... only one specific piece, the Cycles renderer, requires C++17. In fact, there are portions which rely on C++11, and others which need C++14. With so many (nearly independent) moving parts, each with its own needs for C++ standard, it's kind of difficult to use `compiler.cxx_standard` to force a single C++ standard onto the entire source code. Rather, I've been using it as a sort of minimum... Blender uses newer C++ standards in the individual components' CMake files wherever it needs to.

thats not how it works. The setting does not enforce anything on the build, but indicates to macports what 
Capabilities the compiler it picks needs. so you need to in fact give the most recent required standard, and not the oldest. I.e. if any part of the build requires c++17 thats what you must specify, even if other parts only need something older.

> 
> > it would also be useful to know if you in addition blacklist older macports clang versions, ie macports-clang-[5-9].0 which then will force the use of the most recent clang 11 version, if this then means the hack here is no longer required ?
> 
> As far as I'm aware, I'm not blacklisting older macports clangs; I'm only blacklisting Xcode Clang. Could you point out which part of my portfile you are referring to?

i am suggesting you try adding to the blacklist an entry for macports clang, to enforce a more recent version than the default. I.e. try adding

{macports-clang-[5-9].0} 

to the list, which will black list clang 5 to 9, and thus force use of 11.

> 
> > B.t.w. You should also update the regex
> > ```
> > regexp {clang-mp-([0-9]+(?:\.[0-9]+)*)} \
> >     ${configure.cc} result llvm_version
> > ```
> > 
> > 
> > to work properly with macports clang 10 and newer, i.e. where the major version has double digits and there is no minor version, e.g. macports-clang-11.
> 
> I believe that regex should work with clang 10 and newer. The `(?:)*` makes the `.y` portion of version `x.y` optional.

Reply:
>> The entire build doesn't rely on C++17... only one specific piece, the Cycles renderer, requires C++17. In fact, there are portions which rely on C++11, and others which need C++14. With so many (nearly independent) moving parts, each with its own needs for C++ standard, it's kind of difficult to use compiler.cxx_standard to force a single C++ standard onto the entire source code. Rather, I've been using it as a sort of minimum... Blender uses newer C++ standards in the individual components' CMake files wherever it needs to.
> thats not how it works. The setting does not enforce anything on the build, but indicates to macports what capabilities the compiler it picks needs. So you need to in fact give the most recent required standard, and not the oldest. I.e. if any part of the build requires c++17 thats what you must specify, even if other parts only need something older.

Sorry to contradict you on this, I hope you believe me when I say that I'm really not trying to be purposely difficult. But I just tested it, and it looks like `compiler.cxx_standard` is indeed enforcing compile flags during the build. Setting `compiler.cxx_standard 2017` is causing the build to add `-std=c++17` to all compile commands during the build phase. When the line is set to `compiler.cxx_standard 2011` as I had it before, the build is adding `-std=c++11` to all compile commands. Regardless, the build is still running to completion when set to C++17, so I've gone ahead and changed it as requested.

> i am suggesting you try adding to the blacklist an entry for macports clang, to enforce a more recent version than the default. I.e. try adding
> 
> `{macports-clang-[5-9].0}`
> 
> to the list, which will black list clang 5 to 9, and thus force use of 11.

I've added macports clang 5-8 to the blacklist. I've left off version 9, because that is the version of clang being used to compile Blender for its official release binaries, and I would like to keep the possibility of performing a macports build that is "as close to" the official release build as possible.